Source/WebCore/platform/graphics/ImageFrameWorkQueue.cpp

```diff
@@ -0,0 +1,168 @@
+                    LOG(Images, "ImageFrameProvider::%s - %p - url: %s. WorkQueue was recreated at index = %d.", __FUNCTION__, protectedThis.ptr(), protectedFrameProvider->sourceUTF8(), request.index);
+                    return;
+                }
+
+                // The DecodeQueue may have been cleared before the frame was decoded.
+                if (protectedThis->decodeQueue().isEmpty() || protectedThis->decodeQueue().first() != request) {
+                    LOG(Images, "ImageFrameProvider::%s - %p - url: %s. DecodeQueue was cleared at index = %d.", __FUNCTION__, protectedThis.ptr(), protectedFrameProvider->sourceUTF8(), request.index);
+                    return;
+                }
+
+                protectedThis->decodeQueue().removeFirst();
+                protectedFrameProvider->imageFrameDecodeAtIndexHasFinished(request.index, request.subsamplingLevel, request.animatingState, request.options, WTFMove(nativeImage));
+            });
+        }
+
+        // Ensure destruction happens on creation thread.
+        callOnMainThread([protectedThis = WTFMove(protectedThis), protectedWorkQueue = WTFMove(protectedWorkQueue), protectedFrameProvider = WTFMove(protectedFrameProvider)] () mutable { });
+    });
+}
+
+void ImageFrameWorkQueue::dispatch(const Request& request)
+{
+    requestQueue().enqueue(request);
+    decodeQueue().append(request);
+
+    start();
+}
+
+void ImageFrameWorkQueue::stop()
+{
+    for (auto& request : m_decodeQueue) {
+        LOG(Images, "ImageFrameWorkQueue::%s - %p - url: %s. Decoding was cancelled for frame at index = %d.", __FUNCTION__, this, m_frameProvider.sourceUTF8(), request.index);
+        m_frameProvider.clearFrameAtIndex(request.index);
+    }
+
+    if (m_requestQueue) {
+        m_requestQueue->close();
+        m_requestQueue = nullptr;
+    }
+
+    m_decodeQueue.clear();
+    m_commitMap.clear();
+    m_workQueue = nullptr;
+}
+
+void ImageFrameWorkQueue::didCacheNativeImageAtIndex(unsigned index)
+{
+    m_commitMap.set(index + 1, MonotonicTime::now());
+}
+
+void ImageFrameWorkQueue::didDestroyNativeImageAtIndex(unsigned index)
+{
+    m_commitMap.remove(index + 1);
+}
+
+bool ImageFrameWorkQueue::isPendingDecodingAtIndex(unsigned index, SubsamplingLevel subsamplingLevel, const DecodingOptions& options) const
+{
+    auto it = std::find_if(m_decodeQueue.begin(), m_decodeQueue.end(), [index, subsamplingLevel, &options](const Request& request) {
+        return request.index == index && subsamplingLevel >= request.subsamplingLevel && request.options.isCompatibleWith(options);
+    });
+    return it != m_decodeQueue.end();
+}
+
+bool ImageFrameWorkQueue::isPendingDrawingAtIndex(unsigned index) const
+{
+    static constexpr Seconds pendingDrawingInterval = 2_s;
+    if (auto time = m_commitMap.get(index + 1))
+        return MonotonicTime::now() - time < pendingDrawingInterval;
+    return false;
+}
```

Comment:
Is this a new behavior? I couldn't see any reference to a 2s interval in the existing code. What is this for?

Reply:
The logic for destroying the decoded frames in `BitmapImage::destroyDecodedData()` was a little bit cumbersome. We used to destroy all the decoded frames except the first frame and the current frame.  If `destroyAll`  is true we have to consult the `RenderElement::canDestroyDecodedData()` to decide whether to destroy all the decoded frames or not.

In this PR I tried to make it simpler in `ImageFrameProvider::destroyDecodedData()`. The new rule is:
Remove all the decoded frames expect the following two cases:
1. If the frame is pending for drawing, we should not destroy it. We decoded it, cached it but has not drawn it yet. The definition for pending for drawing is in this function.
2. If `destroyAll` is false and the image is animated, the current frame should be cached. It is expensive to decode the next frame of an animated image without having the current frame.

Reply:
OK. That sounds reasonable, but I think any deliberate behavior changes like this really should have been in a separate patch.